crates/polars-io/src/json/mod.rs

```diff
@@ -155,6 +161,101 @@ where
     }
 }
 
+pub trait BatchedWriter<W: Write> {
+    fn write_batch(&mut self, df: &DataFrame) -> PolarsResult<()>;
+    fn finish(&mut self) -> PolarsResult<()>;
+}
+
+pub struct JsonBatchedWriter<W: Write> {
+    writer: W,
+    is_first_row: bool,
+}
+
+impl<W> JsonBatchedWriter<W>
+where
+    W: Write,
+{
+    pub fn new(writer: W) -> Self {
+        JsonBatchedWriter {
+            writer,
+            is_first_row: true,
+        }
+    }
+}
+
+impl<W> BatchedWriter<W> for JsonBatchedWriter<W>
+where
+    W: Write,
+{
+    /// Write a batch to the json writer.
+    ///
+    /// # Panics
+    /// The caller must ensure the chunks in the given [`DataFrame`] are aligned.
+    fn write_batch(&mut self, df: &DataFrame) -> PolarsResult<()> {
+        let fields = df.iter().map(|s| s.field().to_arrow()).collect::<Vec<_>>();
+        let batches = df
+            .iter_chunks()
+            .map(|chunk| Ok(Box::new(chunk_to_struct(chunk, fields.clone())) as ArrayRef));
+        let mut serializer = json::write::Serializer::new(batches, vec![]);
+
+        while let Some(block) = serializer.next()? {
+            if self.is_first_row {
+                self.writer.write_all(&[b'['])?;
+                self.is_first_row = false;
+            } else {
+                self.writer.write_all(&[b','])?;
+            }
+            self.writer.write_all(block)?;
+        }
+        Ok(())
+    }
+
+    /// Writes the footer of the Json file.
+    fn finish(&mut self) -> PolarsResult<()> {
+        self.writer.write_all(&[b']'])?;
+        Ok(())
+    }
+}
+
+pub struct JsonLinesBatchedWriter<W: Write> {
+    writer: W,
+}
+
+impl<W> JsonLinesBatchedWriter<W>
+where
+    W: Write,
+{
+    pub fn new(writer: W) -> Self {
+        JsonLinesBatchedWriter { writer }
+    }
+}
+
+impl<W> BatchedWriter<W> for JsonLinesBatchedWriter<W>
+where
+    W: Write,
+{
+    /// Write a batch to the json writer.
+    ///
+    /// # Panics
+    /// The caller must ensure the chunks in the given [`DataFrame`] are aligned.
+    fn write_batch(&mut self, df: &DataFrame) -> PolarsResult<()> {
+        let fields = df.iter().map(|s| s.field().to_arrow()).collect::<Vec<_>>();
```

Comment:
Personally I would make a class in between that handles all the generic stuff between `JsonWriter` & `JsonLinesBatchedWriter`. In the final classes, I would just implement an `on_block_write` and `on_finish`.

My 2ct :)

Reply:
Yeah, we were thinking of something similar, that's why we implemented the BatchedWriter trait in the json module but we were thinking that it would be better to expose the SinkWriter trait, that is part of the pipeline execution or maybe a struct with some common code as you mentioned, to each of the io modules, so each module implements this interface and we can remove all the implementations currently in file_sink.rs. Of course this will create an extra dependency as the IO modules will now depend of a trait of the execution core, but I think this will make sense since we will only be exposing a interface or a collections of interfaces that need to be implemented by each module. Please let me know your thoughts on this 😄